src/gui/qgsmessagebar.cpp

```diff
@@ -298,6 +298,17 @@ QgsMessageBarItem *QgsMessageBar::pushWidget( QWidget *widget, Qgis::MessageLeve
 
 void QgsMessageBar::pushMessage( const QString &title, const QString &text, Qgis::MessageLevel level, int duration )
 {
+  // refrain from adding more than 20 items in the message bar, avoids flooding (and freezing) of the main window
+  if ( mItems.count() > 20 )
```

Comment:
Instead of blocking new items in this case, would we be better off removing the earliest one?

Reply:
@nyalldawson , I'm not sure. It'd be costlier to do so, and it seems would defeat the flooding protection intent here. Also, for items served with a timeout, it's essentially reset the timeout every time a duplicate item is added (which means that in the case of the auth error, it'd stick there for a looongggg time :) )

Reply:
@nyalldawson , gave this some more thoughts, and adjusted the commit to act as per your suggestion.